This is the exploration notebook meant for learning more about the datasets I have and how I can use them.

In [2]:
import pandas as pd

In [5]:
games_table = pd.read_csv("../data/raw/games.csv")
games_table.head()

,game_id,season,game_type,week,gameday,weekday,gametime,away_team,away_score,home_team,...,wind,away_qb_id,home_qb_id,away_qb_name,home_qb_name,away_coach,home_coach,referee,stadium_id,stadium
0,1999_01_MIN_ATL,1999,REG,1,1999-09-12,Sunday,NaN,MIN,17.0,ATL,...,NaN,00-0003761,00-0002876,Randall Cunningham,Chris Chandler,Dennis Green,Dan Reeves,Gerry Austin,ATL00,Georgia Dome
1,1999_01_KC_CHI,1999,REG,1,1999-09-12,Sunday,NaN,KC,17.0,CHI,...,12.0,00-0006300,00-0010560,Elvis Grbac,Shane Matthews,Gunther Cunningham,Dick Jauron,Phil Luckett,CHI98,Soldier Field
2,1999_01_PIT_CLE,1999,REG,1,1999-09-12,Sunday,NaN,PIT,43.0,CLE,...,12.0,00-0015700,00-0004230,Kordell Stewart,Ty Detmer,Bill Cowher,Chris Palmer,Bob McElwee,CLE00,Cleveland Browns Stadium
3,1999_01_OAK_GB,1999,REG,1,1999-09-12,Sunday,NaN,OAK,24.0,GB,...,10.0,00-0005741,00-0005106,Rich Gannon,Brett Favre,Jon Gruden,Ray Rhodes,Tony Corrente,GNB00,Lambeau Field
4,1999_01_BUF_IND,1999,REG,1,1999-09-12,Sunday,NaN,BUF,14.0,IND,...,NaN,00-0005363,00-0010346,Doug Flutie,Peyton Manning,Wade Phillips,Jim Mora,Ron Blum,IND99,RCA Dome


In [6]:
# Rows are [0] and columns are [1]
game_table_shape = games_table.shape
print(game_table_shape)

(7548, 46)


In [11]:
column_names = games_table.columns
column_names

Index(['game_id', 'season', 'game_type', 'week', 'gameday', 'weekday',
       'gametime', 'away_team', 'away_score', 'home_team', 'home_score',
       'location', 'result', 'total', 'overtime', 'old_game_id', 'gsis',
       'nfl_detail_id', 'pfr', 'pff', 'espn', 'ftn', 'away_rest', 'home_rest',
       'away_moneyline', 'home_moneyline', 'spread_line', 'away_spread_odds',
       'home_spread_odds', 'total_line', 'under_odds', 'over_odds', 'div_game',
       'roof', 'surface', 'temp', 'wind', 'away_qb_id', 'home_qb_id',
       'away_qb_name', 'home_qb_name', 'away_coach', 'home_coach', 'referee',
       'stadium_id', 'stadium'],
      dtype='str')

In [8]:
games_table.info(verbose=True)

<class 'pandas.DataFrame'>
RangeIndex: 7548 entries, 0 to 7547
Data columns (total 46 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   game_id           7548 non-null   str    
 1   season            7548 non-null   int64  
 2   game_type         7548 non-null   str    
 3   week              7548 non-null   int64  
 4   gameday           7548 non-null   str    
 5   weekday           7548 non-null   str    
 6   gametime          7289 non-null   str    
 7   away_team         7548 non-null   str    
 8   away_score        7276 non-null   float64
 9   home_team         7548 non-null   str    
 10  home_score        7276 non-null   float64
 11  location          7548 non-null   str    
 12  result            7276 non-null   float64
 13  total             7276 non-null   float64
 14  overtime          7276 non-null   float64
 15  old_game_id       7548 non-null   int64  
 16  gsis              7276 non-null   float64
 17  nfl_de

In [24]:
games_table_missing = games_table.isna().sum()
games_table_missing[games_table_missing > 0]

gametime             259
away_score           272
home_score           272
result               272
total                272
overtime             272
gsis                 272
nfl_detail_id       7276
pff                 2984
ftn                 5894
away_moneyline      2185
home_moneyline      2185
spread_line          204
away_spread_odds    2185
home_spread_odds    2185
total_line           204
under_odds          2188
over_odds           2188
roof                  43
surface               44
temp                2342
wind                2342
away_qb_id           272
home_qb_id           272
away_qb_name         272
home_qb_name         272
referee              273
dtype: int64

In [10]:
team_table15 = pd.read_csv("../data/raw/stats_team_week_2015.csv")
team_table15

,season,week,team,season_type,opponent_team,completions,attempts,passing_yards,passing_tds,passing_interceptions,...,pat_made,pat_att,pat_missed,pat_blocked,pat_pct,gwfg_made,gwfg_att,gwfg_missed,gwfg_blocked,gwfg_distance
0,2015,1,ARI,REG,NO,19,32,307,3,0,...,4,4,0,0,1.0,0,0,0,0,0
1,2015,1,ATL,REG,PHI,23,34,298,2,2,...,2,2,0,0,1.0,0,0,0,0,0
2,2015,1,BAL,REG,DEN,18,32,117,0,2,...,1,1,0,0,1.0,0,0,0,0,0
3,2015,1,BUF,REG,IND,14,19,195,1,0,...,3,3,0,0,1.0,0,0,0,0,0
4,2015,1,CAR,REG,JAX,18,31,175,1,1,...,2,2,0,0,1.0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
529,2015,20,CAR,POST,ARI,19,28,335,2,1,...,5,5,0,0,1.0,0,0,0,0,0
530,2015,20,DEN,POST,NE,17,32,176,2,0,...,2,2,0,0,1.0,0,0,0,0,0
531,2015,20,NE,POST,DEN,27,56,310,1,2,...,0,1,1,0,0.0,0,0,0,0,0
532,2015,21,CAR,POST,DEN,18,41,265,0,1,...,1,1,0,0,1.0,0,0,0,0,0


In [ ]:
team_table15.columns

Index(['season', 'week', 'team', 'season_type', 'opponent_team', 'completions',
       'attempts', 'passing_yards', 'passing_tds', 'passing_interceptions',
       ...
       'pat_made', 'pat_att', 'pat_missed', 'pat_blocked', 'pat_pct',
       'gwfg_made', 'gwfg_att', 'gwfg_missed', 'gwfg_blocked',
       'gwfg_distance'],
      dtype='str', length=102)

In [23]:
missing_col = team_table15.isna().sum()
missing_col[missing_col > 0]

fg_long            102
fg_pct              80
fg_made_list       102
fg_missed_list     415
fg_blocked_list    510
pat_pct             41
dtype: int64

In [25]:
team_table20 = pd.read_csv('../data/raw/stats_team_week_2020.csv')
team_table20

,season,week,team,season_type,opponent_team,completions,attempts,passing_yards,passing_tds,passing_interceptions,...,pat_made,pat_att,pat_missed,pat_blocked,pat_pct,gwfg_made,gwfg_att,gwfg_missed,gwfg_blocked,gwfg_distance
0,2020,1,ARI,REG,SF,26,40,230,1,1,...,3,3,0,0,1.0,0,0,0,0,0
1,2020,1,ATL,REG,SEA,37,54,450,2,1,...,1,2,1,0,0.5,0,0,0,0,0
2,2020,1,BAL,REG,CLE,21,26,284,3,0,...,5,5,0,0,1.0,0,0,0,0,0
3,2020,1,BUF,REG,NYJ,33,46,312,2,0,...,3,3,0,0,1.0,0,0,0,0,0
4,2020,1,CAR,REG,LV,22,34,269,1,0,...,1,2,1,0,0.5,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
533,2020,20,GB,POST,TB,33,48,346,3,1,...,2,2,0,0,1.0,0,0,0,0,0
534,2020,20,KC,POST,BUF,29,38,325,3,0,...,5,5,0,0,1.0,0,0,0,0,0
535,2020,20,TB,POST,GB,20,36,280,3,3,...,4,4,0,0,1.0,0,0,0,0,0
536,2020,21,KC,POST,TB,26,49,270,0,2,...,0,0,0,0,NaN,0,0,0,0,0


In [26]:
team_table24 = pd.read_csv('../data/raw/stats_team_week_2024.csv')
team_table24

,season,week,team,season_type,opponent_team,completions,attempts,passing_yards,passing_tds,passing_interceptions,...,pat_made,pat_att,pat_missed,pat_blocked,pat_pct,gwfg_made,gwfg_att,gwfg_missed,gwfg_blocked,gwfg_distance
0,2024,1,ARI,REG,BUF,21,31,162,1,0,...,2,2,0,0,1.0,0,0,0,0,0
1,2024,1,ATL,REG,PIT,16,26,155,1,2,...,1,1,0,0,1.0,0,0,0,0,0
2,2024,1,BAL,REG,KC,26,41,273,1,0,...,2,2,0,0,1.0,0,0,0,0,0
3,2024,1,BUF,REG,ARI,18,23,232,2,0,...,4,4,0,0,1.0,0,0,0,0,0
4,2024,1,CAR,REG,NO,13,31,161,0,2,...,1,1,0,0,1.0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
565,2024,21,KC,POST,BUF,18,26,245,1,0,...,3,3,0,0,1.0,0,0,0,0,0
566,2024,21,PHI,POST,WAS,20,28,246,1,0,...,7,7,0,0,1.0,0,0,0,0,0
567,2024,21,WAS,POST,PHI,30,49,278,1,1,...,0,0,0,0,NaN,0,0,0,0,0
568,2024,22,KC,POST,PHI,21,32,257,3,2,...,0,0,0,0,NaN,0,0,0,0,0


In [27]:
# Loading rest of the team tables
team_table16 = pd.read_csv('../data/raw/stats_team_week_2016.csv')
team_table17 = pd.read_csv('../data/raw/stats_team_week_2017.csv')
team_table18 = pd.read_csv('../data/raw/stats_team_week_2018.csv')
team_table19 = pd.read_csv('../data/raw/stats_team_week_2019.csv')
team_table21 = pd.read_csv('../data/raw/stats_team_week_2021.csv')
team_table22 = pd.read_csv('../data/raw/stats_team_week_2022.csv')
team_table23 = pd.read_csv('../data/raw/stats_team_week_2023.csv')

In [30]:
# Checking for duplicate rows in all tables
games_table.duplicated().sum()
team_table15.duplicated().sum()
team_table16.duplicated().sum()
team_table17.duplicated().sum()
team_table18.duplicated().sum()
team_table19.duplicated().sum()
team_table20.duplicated().sum()
team_table21.duplicated().sum()
team_table22.duplicated().sum()
team_table23.duplicated().sum()
team_table24.duplicated().sum()

np.int64(0)

Observations of games.csv dataset:
Shape: 7548 rows x 46 columns
Each row represents one game
0 duplicate rows in any of the csv files
Possible useful columns are:
    Outcomes:
        - away_score
        - home_score
    Situational:
        - season
        - game_type
        - week
        - away_team
        - home_team
        - home_rest
        - away_rest

Observations of stats_team_week_2015.csv
Shape: 534 rows x 102 columns
Each row represents team stats on all sides of the ball for one week
Different types of season_type
Structure is the same between the other stats_team_week csv files
0 duplicate rows in any of the csv files
Possible useful columns are:
    Offensive:
        - passing_yards
        - passing_tds
        - passing_interceptions
        - rushing_yards
        - rushing_tds
        - rushing_fumbles_lost
    Defensive:
        - def_tackles_for_loss
        - def_fumbles_forced
        - def_sacks
        - def_interceptions
        - def_qb_hits
    Special Teams:
        - kickoff_return_avg(will need to derive this)
        - fg_pct
        - pat_pct
    Situational:
        - week
        - team
        - season_type
        - penalties
        - penalty_yards
Target variable: Whether the team represented in the row won the game (Win/Loss)

games and the stats_team_week csv files are matched and able to be combined

Potential Data Leakage columns(both):
    - away_score
    - home_score
    - Any weekly team statistic from the game being predicted
        - These statistics must be converted to season-to-date-averages in order to be used in the model
    - penalties
    - penalty_yards

